### ISSUE TITLE
[MRESOLVER-123] Provide a global locking sync context by default

### ISSUE THREAD

**Issue by jira-importer** [apache/maven-resolver#919]
**[Michael Osipov](https://issues.apache.org/jira/secure/ViewProfile.jspa?name=michael-o)** opened **[MRESOLVER-123](https://issues.apache.org/jira/browse/MRESOLVER-123?redirect=false)** and commented

This is an umbrella ticket for a long standing issue with Maven Resolver: Our concurrency support is mediocre in a way that if two or more threads try to download the same file and fail to queue those write actions nicely. The problem is that The `SyncContext` and the its factory provided by Maven Resolver does not employ any locking at all. As layed out in detail in MRESOLVER-114 we need striped read write locks on artifacts and its metadata. This issue shall track progress on it. Even Takari Concurrent Repository extension does not help because it is only intended to synchronize concurrent access by multple JVMs and not threads.

This improvement will provide solely a global locking sync context which will work in **single** JVM. It is a non-goal to make it work with mulitple JVMs. A downside of this solution is that is coarse, possible degregation of performance for the sake of stability.

---

**Affects:** 1.4.2

**Attachments:**
- [checksum-error-debug.log](https://issues.apache.org/jira/secure/attachment/13006740/checksum-error-debug.log) (_858.16 kB_)
- [mvn-debug-1.4.3-123.txt.gz](https://issues.apache.org/jira/secure/attachment/13007139/mvn-debug-1.4.3-123.txt.gz) (_6.55 MB_)

**Issue Links:**
- [MRESOLVER-25](https://issues.apache.org/jira/browse/MRESOLVER-25) Resume support is broken under high concurrency
 (_**"fixes"**_)
- [MRESOLVER-114](https://issues.apache.org/jira/browse/MRESOLVER-114) ArtifactNotFoundExceptions when building in parallel
 (_**"fixes"**_)
- [MRESOLVER-178](https://issues.apache.org/jira/browse/MRESOLVER-178) Introduce a simple inter-process  SyncContext

- [MRESOLVER-131](https://issues.apache.org/jira/browse/MRESOLVER-131) Introduce a Redisson-based SyncContextFactory

- [MNG-2802](https://issues.apache.org/jira/browse/MNG-2802) Concurrent-safe access to local Maven repository

- [MRESOLVER-130](https://issues.apache.org/jira/browse/MRESOLVER-130) Move GlobalSyncContextFactory to a separate module
 (_**"is superceded by"**_)

**Remote Links:**
- [GitHub Pull Request #65
](https://github.com/apache/maven-resolver/pull/65)

**Backported to:** [1.6.0](https://github.com/apache/maven-resolver/milestone/47?closed=1), [1.5.0](https://github.com/apache/maven-resolver/milestone/46?closed=1)

0 votes, 8 watchers
